File: cli/node/src/runtime/state.ts
import { AsyncLocalStorage } from 'node:async_hooks';
import fs from 'node:fs';

/** 单次调用的终端状态；仅运行时边界允许更新。 */
export interface InvocationState {
  agentMode: boolean;
  currentCommand: string;
  pendingNotice: string;
}
const invocations = new AsyncLocalStorage<InvocationState>();

/**
 * 创建独立上下文，避免多次执行或并发命令相互污染。
 * @returns 独立的调用状态对象。
 */
export function createInvocationState(): InvocationState {
  return { agentMode: false, currentCommand: '', pendingNotice: '' };
}
/**
 * 在调用范围内装配终端副作用依赖。
 * @param state - 本次调用独立持有的状态。
 * @param action - 需要在指定上下文内执行的函数。
 * @returns 在指定上下文中执行的函数返回值。
 */
export function withInvocation<T>(state: InvocationState, action: () => T): T {
  return invocations.run(state, action);
}
/**
 * 为测试或嵌入式调用绑定当前执行上下文。
 * @param state - 本次调用独立持有的状态。
 */
export function enterInvocation(state: InvocationState): void {
  invocations.enterWith(state);
}
/**
 * 读取本次调用状态；缺少上下文时明确报错。
 * @returns 当前异步调用上下文中的状态。
 */
function current(): InvocationState {
  const state = invocations.getStore();
  if (!state) throw new Error('CLI invocation context is missing. Use withInvocation().');
  return state;
}
/**
 * 读取本次调用的代理输出开关。
 * @returns 当前是否使用代理输出模式。
 */
export function isAgentMode(): boolean {
  return current().agentMode;
}
/**
 * 设置本次调用的代理输出开关。
 * @param value - 是否采用代理输出模式。
 */
export function setAgentMode(value: boolean): void {
  current().agentMode = value;
}
/**
 * 读取当前命令名称。
 * @returns 当前命令名称。
 */
export function getCurrentCommand(): string {
  return current().currentCommand;
}
/**
 * 记录当前命令名称供输出信封使用。
 * @param name - 调用方或命令名称。
 */
export function setCurrentCommand(name: string): void {
  current().currentCommand = name;
}
/**
 * 后端提示每条命令只输出一次，不随请求数量重复。
 * @param notice - 等待输出的平台通知。
 */
export function captureNotice(notice: string | null | undefined): void {
  if (notice) current().pendingNotice = notice;
}
/**
 * 取出待展示通知并清空本次调用中的缓存。
 * @returns 本次取出的通知，未设置时为空字符串。
 */
export function takeNotice(): string {
  const state = current();
  const notice = state.pendingNotice;
  state.pendingNotice = '';
  return notice;
}
/**
 * 只读实际管道或重定向文件，避免 /dev/null 和 socket 的 EAGAIN。
 * @returns 当前是否允许并检测到文件或管道输入。
 */
export function stdinIsPiped(): boolean {
  if (isAgentMode()) return false;
  const stat = fs.fstatSync(0);
  return stat.isFIFO() || stat.isFile();
}

File: cli/python/src/memgo_cli/runtime/state.py
"""用 ContextVar 隔离命令状态，避免同进程多次调用相互污染。"""

from __future__ import annotations

from collections.abc import Callable
from contextlib import ExitStack
from contextvars import ContextVar, Token
from dataclasses import dataclass, replace

from rich.console import Console


@dataclass(frozen=True)
class InvocationState:
    """单次调用的输出、身份和资源上下文。"""

    agent_mode: bool = False
    current_command: str = ""
    pending_notice: str = ""
    validated_email: str | None = None
    resources: ExitStack | None = None


# 默认值是冻结模型且不持有资源；所有更新均替换对象。
_state: ContextVar[InvocationState] = ContextVar("memgo_invocation", default=InvocationState())  # noqa: B039


def start_invocation(agent_mode: bool) -> Token[InvocationState]:
    """建立独立上下文，返回用于恢复外层状态的令牌。"""
    return _state.set(InvocationState(agent_mode=agent_mode, resources=ExitStack()))


def finish_invocation(token: Token[InvocationState]) -> None:
    """关闭连接并展示待处理通知，最后恢复外层调用状态。"""
    state = _state.get()
    try:
        if state.resources is not None:
            state.resources.close()
        if state.pending_notice and not state.agent_mode:
            Console(stderr=True).print(f"\n[yellow]🔔 {state.pending_notice}[/yellow]\n")
    finally:
        _state.reset(token)


def register_cleanup(action: Callable[[], None]) -> None:
    """为本次调用登记资源关闭操作。"""
    resources = _state.get().resources
    if resources is None:
        raise RuntimeError("CLI invocation context is missing")
    resources.callback(action)


def is_agent_mode() -> bool:
    """读取本次调用的代理输出开关。"""
    return _state.get().agent_mode


def set_agent_mode(val: bool) -> None:
    """替换本次调用的代理输出开关。"""
    _state.set(replace(_state.get(), agent_mode=val))


def get_current_command() -> str:
    """读取当前命令名。"""
    return _state.get().current_command


def set_current_command(name: str) -> None:
    """记录当前命令名供错误信封使用。"""
    _state.set(replace(_state.get(), current_command=name))


def capture_notice(notice: str | None) -> None:
    """缓存本次调用的最新非空平台通知。"""
    if notice:
        _state.set(replace(_state.get(), pending_notice=notice))


def take_notice() -> str:
    """取出平台通知并清除缓存。"""
    state = _state.get()
    _state.set(replace(state, pending_notice=""))
    return state.pending_notice


def validated_email() -> str | None:
    """读取当前调用已验证的邮箱。"""
    return _state.get().validated_email


def set_validated_email(email: str) -> None:
    """保存当前调用已验证的邮箱。"""
    _state.set(replace(_state.get(), validated_email=email))


def caller_type() -> str:
    """提供后端请求使用的调用方类型。"""
    return "agent" if is_agent_mode() else "user"
